src/lib/png/png.c

```diff
@@ -0,0 +1,976 @@
+			} else if (code == 16) /* repeat previous */ {
+				replength = 3; /* read in the 2 bits that indicate repeat length (3-6) */
+
+				if (i == 0) {
+					err = -1; /* can't repeat previous if i is 0 */
+					break;
+				}
+
+				if ((*bp + 2) > in_bitlen) {
+					err = -1; /* error, bit pointer jumps past memory */
+					break;
+				}
+				replength += read_bits_from_stream(bp, in, 2);
+
+				if (i < HLIT + 1) {
+					value = bitlen_ll[i - 1];
+				} else {
+					value = bitlen_d[i - HLIT - 1];
+				}
+				/* repeat this value in the next lengths */
+				for (n = 0; n < replength; n++) {
+					if (i >= HLIT + HDIST) {
+						err = -1; /* error: i is larger than the amount of codes */
+						break;
+					}
+					if (i < HLIT) {
+						bitlen_ll[i] = value;
+					} else {
+						bitlen_d[i - HLIT] = value;
+					}
+					i++;
+				}
+			} else if (code == 17) /* repeat "0" 3-10 times */ {
+				replength = 3; /* read in the bits that indicate repeat length */
+
+				if ((*bp + 3) > in_bitlen) {
+					err = -1; /* error, bit pointer jumps past memory */
+					break;
+				}
+
+				replength += read_bits_from_stream(bp, in, 3);
+
+				/* repeat this value in the next lengths */
+				for (n = 0; n < replength; n++) {
+					if (i >= HLIT + HDIST) {
+						return -1; /* error: i is larger than the amount of codes */
+					}
+
+					if (i < HLIT) {
+						bitlen_ll[i] = 0;
+					} else {
+						bitlen_d[i - HLIT] = 0;
+					}
+					i++;
+				}
+			} else if (code == 18) /* repeat "0" 11-138 times */ {
+				replength = 11; /* read in the bits that indicate repeat length */
+				if ((*bp + 7) > in_bitlen) {
+					err = -1; /* error, bit pointer jumps past memory */
+					break;
+				}
+				replength += read_bits_from_stream(bp, in, 7);
+
+				/* repeat this value in the next lengths */
+				for (n = 0; n < replength; n++) {
+					if (i >= HLIT + HDIST) {
+						err = -1; /* error: i is larger than the amount of codes */
+						break;
+					}
+					if (i < HLIT) {
+						bitlen_ll[i] = 0;
+					} else {
+						bitlen_d[i - HLIT] = 0;
+					}
+					i++;
+				}
+			} else {
+				if (code == -1) {
+					err = -1; /* error in decodesymbol fun */
+					break;
+				}
+			}
+		}
+		if (err) {
+			break;
+		}
+
+		if (bitlen_ll[256] == 0) {
+			err = -1; /* the length of the end code 256 must be larger than 0 */
+			break;
+		}
+
+		/* generate the code trees */
+		err = make_tree_first_step(ll_tree, bitlen_ll, NUMBER_OF_CODE_SYMBOLS, 15);
+
+		if (err) {
+			break;
+		}
+
+		err = make_tree_first_step(dist_tree, bitlen_d, NUMBER_OF_DISTANCE_SYMBOLS, 15);
+
+		break; /* end of while (!err) */
+	}
+	HuffmanTree_cleanup(&cl_tree);
+	return err;
+}
+
+/* get the dist_tree with fixed huffman */
+static int build_fixed_dist_tree(HuffmanTree *tree) {
+	int i, err = 0;
+	int bitlen[NUMBER_OF_DISTANCE_SYMBOLS];
+
+	for (i = 0; i != NUMBER_OF_DISTANCE_SYMBOLS; i++) {
+		bitlen[i] = 5;
+	}
+
+	err = make_tree_first_step(tree, bitlen, NUMBER_OF_DISTANCE_SYMBOLS, 15);
```

Comment:
You can simplify this code by returning function value directly:

`return make_tree_...`

Reply:
done